Change all instance of throwing strings to throw specific error classes.

chingjun opened the pull request:
This is a followup to one of the feedbacks in #95738.

I changed all the cases where we throw a string in flutter_tools to throw specific error classes instead, and tried my best to choose the correct class to throw.

Please do correct me if I am choosing the incorrect class.

Also for the case in Daemon, I created a new `DaemonError` class to avoid changing the error message sent to the client. By default, an `Exception` class would add a prefix `"Exception: "` to the message, which might be backward incompatible with existing clients.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

chingjun committed:
Change all instance of throwing strings to throw specific error classes.

zanderso commented on packages/flutter_tools/lib/src/commands/daemon.dart:
```diff
@@ -1556,3 +1556,13 @@ class DebounceOperationQueue<T, K> {
 }
+
+/// Specialized exception for returning errors to the daemon client.
+class DaemonError implements Exception {
```
IIRC we try to make it clear that a class is an `Exception` in the name so that it will be obvious that it will be caught by catch blocks that handle the `Exception` type.

```suggestion
class DaemonException implements Exception {
```

christopherfujino replied:
+1

chingjun replied:
Done

zanderso commented on packages/flutter_tools/lib/src/commands/update_packages.dart:
```diff
@@ -774,7 +774,7 @@ class PubspecYaml {
               // twice doesn't make sense.
               if (masterDependencies.containsKey(dependency.name)) {
-                throw '$filename contains two dependencies on ${dependency.name}.';
+                throw Exception('$filename contains two dependencies on ${dependency.name}.');
```
If this happens in a file under user control, it should also be a `throwToolExit`.

zanderso commented on packages/flutter_tools/lib/src/commands/update_packages.dart:
```diff
@@ -1306,7 +1306,7 @@ class PubspecDependency extends PubspecLine {
       return false;
     } else {
-      throw 'Could not parse additional details for dependency $name; line was: "$line"';
+      throw Exception('Could not parse additional details for dependency $name; line was: "$line"');
```
ditto

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -536,16 +536,16 @@ class FuchsiaDevice extends Device {
   Future<void> takeScreenshot(File outputFile) async {
     if (outputFile.basename.split('.').last != 'ppm') {
-      throw '${outputFile.path} must be a .ppm file';
+      throw Exception('${outputFile.path} must be a .ppm file');
```
If the call site of `takeScreenshot` is catching a `String` then it would need to be updated. What do the `takeScreenshot` implementations on other devices do?

chingjun replied:
The call site is actually catching an `Exception` (https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/screenshot.dart#L132), and I don't think this is using a String to bypass that catch block.

Other platforms don't have as much code in `takeScreenshot`, they either throw `UnimplementedError` or call `ProcessManager` to actually take screenshots

zanderso commented on packages/flutter_tools/lib/src/resident_runner.dart:
```diff
@@ -633,7 +633,7 @@ abstract class ResidentHandlers {
   Future<OperationResult> restart({ bool fullRestart = false, bool pause = false, String reason }) {
     final String mode = isRunningProfile ? 'profile' :isRunningRelease ? 'release' : 'this';
-    throw '${fullRestart ? 'Restart' : 'Reload'} is not supported in $mode mode';
+    throw UnsupportedError('${fullRestart ? 'Restart' : 'Reload'} is not supported in $mode mode');
```
This is another case where checking what is happening at the call site is a good idea. `UnsupportedError` will cause the tool to crash, whereas the call site might be trying to emit the error string and continuing.

chingjun replied:
I didn't realize the distinction between Error and Exception. I've changed this to `Exception`.

Comment about checking the callsite in another comment below.

zanderso commented on packages/flutter_tools/lib/src/resident_runner.dart:
```diff
@@ -1280,7 +1280,7 @@ abstract class ResidentRunner extends ResidentHandlers {
   }) async {
     if (!debuggingOptions.debuggingEnabled) {
-      throw 'The service protocol is not enabled.';
+      throw StateError('The service protocol is not enabled.');
```
ditto

zanderso commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -219,7 +219,7 @@ class HotRunner extends ResidentRunner {
       }
     }
-    throw 'Failed to compile $expression';
+    throw Exception('Failed to compile $expression');
```
ditto

zanderso reviewed (commented):
I'd suggest trimming this PR down to just the changes in `daemon.dart` and figuring out how to test them.

chingjun committed:
Turn on only_throw_errors lint and fix everything that violates this.

chingjun commented:
Thanks for the review!

> I'd suggest trimming this PR down to just the changes in daemon.dart and figuring out how to test them.

Can you elaborate on what kind of tests do you think we need? We currently have some daemon tests in `commands.shard/hermetic/daemon_test.dart` that test some of its input/output behavior.

If you mean we should make sure that we never throw Strings, then I think we should re-enable the [`only_throw_errors` lint](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/analysis_options.yaml#L10), which is disabled "for historical reasons". So that's what I did :) Also changed all the tests that used to throw `String` to throw real error types.

I'm still keeping everything in this PR, do let me know if you still prefer that I split it into multiple smaller ones.

And just FYI in case there is miscommunication, I still intend to send other PRs to address your other feedbacks in #95738.

---

Regarding your comment about checking call sites that catches String, I searched in the whole package and found that there are only 3 places (outside of test/) that catches String explicitly.

Two in [update_packages.dart](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/update_packages.dart#L255) which catches the String and throwToolExit, so I just removed the try block and changed the throw to throwToolExit directly.

The last one is in [run_cold.dart](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/run_cold.dart#L90). I can't think of a good reason that we need to catch a String here, especially since we don't do the same in run_hot.dart. I think the original intention is to do some cleanup if anything fails. I tried looking around and can't find anywhere that is doing useful cleanup after app failed to start, nor there are any places that currently throws `String` that need special care. I changed this to catch a generic `Exception` instead for a generic cleanup.

zanderso reviewed (approved):
Cool. Thanks for the cleanup!

christopherfujino commented on packages/flutter_tools/analysis_options.yaml:
```diff
@@ -7,7 +7,7 @@ linter:
     library_private_types_in_public_api: false # Tool does not have any public API.
     no_runtimeType_toString: false # We use runtimeType for debugging in the tool.
-    only_throw_errors: false # For historical reasons we throw strings a lot.
+    only_throw_errors: true
```
nice!

christopherfujino reviewed (approved):
thanks for this, LGTM
